core/services/vrf/vrfcommon/utils.go

```diff
@@ -0,0 +1,66 @@
+package vrfcommon
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+
+	txmgrcommon "github.com/smartcontractkit/chainlink/v2/common/txmgr"
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/txmgr"
+)
+
+type respCountEntry struct {
+	RequestID string
+	Count     int
+}
+
+func GetRespCounts(ctx context.Context, txm txmgr.TxManager, chainID *big.Int, confirmedBlockNum int64) (
+	[]respCountEntry,
+	error,
+) {
+	counts := []respCountEntry{}
+	metaField := "RequestID"
+	states := []txmgrtypes.TxState{txmgrcommon.TxUnconfirmed, txmgrcommon.TxUnstarted, txmgrcommon.TxInProgress}
+	// Search for txes with a non-null meta field in the provided states
+	unconfirmedTxes, err := txm.FindTxesWithMetaFieldByStates(ctx, metaField, states, chainID)
+	if err != nil {
+		return nil, errors.Wrap(err, "getRespCounts failed due to error in FindTxesWithMetaFieldByStates")
+	}
+	// Fetch completed transactions only as far back as the given cutoffBlockNumber. This avoids
+	// a table scan of the whole table, which could be large if it is unpruned.
+	var confirmedTxes []*txmgr.Tx
+	confirmedTxes, err = txm.FindTxesWithMetaFieldByReceiptBlockNum(ctx, metaField, confirmedBlockNum, chainID)
```

Comment:
One difference I am noticing about this implementation is that FindTxesWithMetaFieldByStates and FindTxesWithMetaFieldByReceiptBlockNum are potentially done in different database transactions (correct me if I'm wrong). I think this can result in issues because the TXM may potentially move stuff from unconfirmed -> confirmed in the meantime. Is it possible to perform these in the same DB tx to avoid this issue and remain backwards compatible?

Reply:
That's a good point. I don't think we want to expose pg opts to allow an atomic transaction here so I may need to add some logic to filter the transactions. I can dedupe the list to ensure that an unconfirmed transaction isn't picked up by the confirmed query after state change to avoid double counting. Would there be any scenario we would want to exclude a tx altogether? Is it possible for a tx to be confirmed but have a receipt outside the confirmed block num which wouldn't be counted under the current queries? Is that even something we would want to exclude?

Reply:
I see 3 race scenarios here:
1. unconfirmed -> confirmed (older than finalization) - tx is in first query but not in second query. nothing to dedupe
2. unconfirmed -> confirmed (more recent than finalization) - tx is in first query and in second query. Dedupe using `tx_id`
3. unconfirmed -> confirmed_missing_receipt - tx is in first query but not in second query. nothing to dedupe

Any other cases?

Reply:
Are there any scenarios, I'd need to remove a tx entirely from the count like in case 1 or 3? Or is it alright if they stay in the list because it may or may not have been picked up in the existing query depending when it was executed?

Reply:
IIUC, in the original atomic query, nothing is removed from the result of the 2 queries (i.e. `unconfirmedQuery` and `confirmedQuery` are appended). So to maintain parity, I don't think any tx needs to be completely removed from the `unconfirmedQuery`. Some txs just need to be deduped only when the `tx` appears in both `unconfirmedQuery` and `confirmedQuery`

Reply:
Ok I understood it as the same. I'll add deduping logic to avoid double counting